modules/core/include/opencv2/core/hal/cvtype.h

```diff
@@ -0,0 +1,258 @@
+//     and/or other materials provided with the distribution.
+//
+//   * The name of the copyright holders may not be used to endorse or promote products
+//     derived from this software without specific prior written permission.
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#ifndef OPENCV_CORE_HAL_CVTYPE_H
+#define OPENCV_CORE_HAL_CVTYPE_H
+
+//! @name OpenCV4 Data types
+//! @sa cv::Types
+//! @{
+
+//ElemType register specifications
+#define CV_DEPTH_SHIFT            0
+#define CV_CN_SHIFT               5
+#define CV_CN_LEN                 7
+#define CV_CN_EXP_LEN             3
+#define CV_CN_BASE_LEN            (CV_CN_LEN - CV_CN_EXP_LEN)
+#define CV_DEPTH_LEN              (CV_CN_SHIFT - CV_DEPTH_SHIFT)
+
+//ElemType exponent power limit (private)
+#define __CV_CN_LIMIT_EXP(n)      (1 << (CV_CN_BASE_LEN + n))
+
+//ElemType register masks (private)
+#define CV_SANITY_DEPTH_MASK      ((1 << CV_DEPTH_LEN  ) - 1)
+#define CV_SANITY_CN_MASK         ((1 << CV_CN_LEN     ) - 1)
+#define CV_SANITY_CN_EXP_MASK     ((1 << CV_CN_EXP_LEN ) - 1)
+#define CV_SANITY_CN_BASE_MASK    (__CV_CN_LIMIT_EXP(0)  - 1)
+#define CV_MAT_DEPTH_MASK         (CV_SANITY_DEPTH_MASK << CV_DEPTH_SHIFT)
+#define CV_MAT_CN_MASK            (CV_SANITY_CN_MASK    << CV_CN_SHIFT   )
+#define CV_MAT_TYPE_MASK          (CV_MAT_CN_MASK | CV_MAT_DEPTH_MASK)
+
+//ElemType register limits (public)
+#define CV_DEPTH_MAX              (1 << CV_DEPTH_LEN)
+#define CV_CN_MAX                 __CV_CN_LIMIT_EXP(CV_SANITY_CN_EXP_MASK) /*(1 << CV_CN_LEN )*/
+
+//ElemType register deflate/inflate operations (private)
+#define __CV_CN_MAKE_EXPONENT(cn) ((cn) <= __CV_CN_LIMIT_EXP(0) ? 0 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(1) ? 1 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(2) ? 2 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(3) ? 3 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(4) ? 4 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(5) ? 5 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(6) ? 6 : 7)))))))
```

Comment:
Are there any **real cases** for requirement of 2048 "floating point" channels?

Reply:
It is rather a design outcome! `(16 << 5) = 512` was the previous limit, and `5 = 101b` already reserves 3bits, which can go up to `7 = 111b`, thus channels can be `(16 << 7) = 2048` by design.